feat(a380/pfd): Add ARS indication on speed tape

BravoMike99 opened the pull request:
<!-- ⚠⚠ Do not delete this pull request template! ⚠⚠ -->
<!-- Pull requests that do not follow this template are likely to be ignored. -->

<!-- Add the issues this PR fixes here. If no issues are related to this PR, then this line can be removed. -->
<!-- Add further issues with a full "Fixes #[issue_no]" line to ensure GitHub closes each one when the PR is merged. -->
Fixes #[issue_no]

## Summary of Changes
<!-- Please provide a summary of changes for this pull request, ensuring all changes are explained. -->
Adds ARS speed indication on PFD speed tape.
Updates target speed bug
## Screenshots (if necessary)
<!-- If your PR includes visual changes, screenshots from before and after your change should always be included. -->
<!-- Please make your best efforts to provide useful before and after screenshots. They should match camera angle, zoom, size, time of day, etc. -->
![image](https://github.com/user-attachments/assets/0fa1991d-974d-4490-a8fb-864046ca8996)
![image](https://github.com/user-attachments/assets/1499f435-3e70-4ad1-a529-bcd082882956)

## References
<!-- You should be making changes based on some kind of a reference (manuals, videos, IRL photos). P3D/xplane/fsx references will only be accepted if we believe that one cannot reasonably obtain a better source. Please post screenshots of the references you used. Ask around in the discord for how to find references for what you are working on. Exceptions will probably be made for IRL A320 pilots and engineers. -->
FCOM & https://youtu.be/yGlS2QBuX7s?t=1664
<!-- If you are making a pull request related to the MCDU, please make sure you are ONLY referencing the Honeywell Pegasus Step 1A (Rev 0), 2009 edition manual. -->
<!-- If you do not have this manual, please ask on our discord for assistance -->

## Additional context
<!-- Add any other context about the pull request here. -->

<!-- You may optionally provide your discord username, so that we may contact you directly about the issue. -->
Discord username (if different from GitHub):
bruno_pt99
## Testing instructions
A32NX regression testing, Perform a full flight:
Check that indications on lower SD page show valid data.
Check that normal law indications show on PFD (green " on top of the attitude indicator) and that the sidestick indication is available after one engine running.
No erroneous ecam/PFD messages or GPWS warnings are triggered during flight.
<!-- Detail how this PR should be tested by QA. Try to list important items that need checking, either directly changed by this PR or that could be affected by it -->
<!-- DO NOT DELETE THIS -->
## How to download the PR for QA

Every new commit to this PR will cause new A32NX and A380X artifacts to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, find and click on the **PR Build** tab
1. Click on either **flybywire-aircraft-a320-neo** or **flybywire-aircraft-a380-842** download link at the bottom of the page

flogross89 commented:
Nice one! Just to understand the behavior: After the a/c passed 212kt with a positive speed trend, the ARS will become active and move the VFE to 263kt?

BravoMike99 commented:
> Nice one! Just to understand the behavior: After the a/c passed 212kt with a positive speed trend, the ARS will become active and move the VFE to 263kt? As a consequence (and that's why I'm asking), as soon as the current speed reference line passes 212kt, the small green ARS line should disappear and the red barber pole should move to 263kt, correct?

Yes exactly.

flogross89 commented on fbw-a380x/src/systems/instruments/src/PFD/SpeedIndicator.tsx:
```diff
@@ -957,6 +957,59 @@ class V1Offtape extends DisplayComponent<{ bus: EventBus }> {
+class ArsBar extends DisplayComponent<{ bus: ArincEventBus }> {
+  private arsPath = Subject.create<string>('');
+
+  private arsVisiblity = Subject.create<string>('hidden');
```
For our newer components, we try to move towards working with ConsumerSubjects and MappedSubjects, instead of using setXxx() functions called after subject updates.
Could you try implementing this? 
Examples are in e.g. https://github.com/flybywiresim/aircraft/blob/master/fbw-a380x/src/systems/instruments/src/EWD/elements/EGT.tsx#L25

For example:
```
private readonly slatsFlapsStatus = Arinc429ConsumerSubject.create(this.sub.on('slatsFlapsStatus').whenChanged());
private readonly conf1FSelected = this.slatsFlapsStatus.map((s) => s.getBitValueOr(28, false) && s.getBitValue(29) && s.getBitValue(18) && !s.getBitValue(26));
private readonly arsVisibility = this.conf1FSelected.map((c) => c ? 'visible' : 'hidden');
```

and so on, feel free to DM me for questions

tracernz replied:
Typescript can also infer the type just fine here, so you don't need to specify it with <string>.

tracernz commented on fbw-a380x/src/systems/instruments/src/PFD/SpeedIndicator.tsx:
```diff
@@ -957,6 +957,59 @@ class V1Offtape extends DisplayComponent<{ bus: EventBus }> {
+
+  private conf1FSelected = false;
+
+  private readonly ARS1FSpeed = 212;
+
+  private readonly Conf1FVFE = 222;
```
These should probably be `private static readonly ARS_1_F_SPEED`. If they're instance members they should be camel case rather than pascal case.

tracernz commented on fbw-a380x/src/systems/instruments/src/PFD/SpeedIndicator.tsx:
```diff
@@ -957,6 +957,59 @@ class V1Offtape extends DisplayComponent<{ bus: EventBus }> {
+    if (this.conf1FSelected) {
+      this.arsVisiblity.set('visible');
+      const arsPos = ((this.airSpeed.value - this.ARS1FSpeed) * DistanceSpacing) / ValueSpacing + 80.818;
+      this.arsPath.set(`m19.031 ${arsPos}h 1.9748v${this.offset}`);
```
It would be much more efficient to transform this rather than changing the path constantly, but I think we really have that problem all over the PFD so I guess it's okay for now.

BravoMike99 commented:
Thanks both for the reviews. I changed the subject logic with floridude's help and changed the vertical offset to use translate3D. I did, however, need to apply one patch provided by holland from the ATA 21 failures PR per floridude suggestion as the getBitValue methods are not exposed in the arinc interface currently.

flogross89 reviewed (approved):
LGTM now :)

tracernz commented on fbw-common/src/systems/shared/src/arinc429.ts:
```diff
@@ -154,14 +160,24 @@ export class Arinc429Register implements Arinc429WordData {
     return ((this.value >> (bit - 1)) & 1) !== 0;
   }
 
+  /** @deprecated use {@link getBitValueOr} */
```
What's the reason for this change? `bitValue` and `bitValueOr` match up with `value` and `valueOr`. These new functions you've introduced do not match up with the rest of the API.

BravoMike99 replied:
Initially I was trying to use Arinc429ConsumerSubjects on the PFD side and as the bitValueOr methods weren't being exposed in the interface I picked the changes from here https://github.com/flybywiresim/aircraft/pull/8883#discussion_r1760085768. In any case, I've changed it so it all uses btValue & bitValueOr & removed the get functions so its consistent across the board.

tylerclt commented:
<!-- !! DO NOT REMOVE THIS LINE !! - QUALITY ASSURANCE REPORT -->
<!-- Refer to the https://docs.flybywiresim.com/dev-corner/qa-process/#reporting-on-tests for details about this template -->

<!-- Please use one or the other depending on your Discord Role -->
Quality Assurance Tester/Trainee Report

Discord Username: tylerb780 <!-- Use your username, not your display name! -->
Object of testing : #8922  <!-- PR Number -->
Aircraft: A32NX <!-- Choose only one -->
Tier of Testing: 2
Date: 22/09/2024 <!-- Date of test in DD/MM/YYYY format -->

Testing Process: 
1. Loaded into the A32NX at KLAS for a flight to KSAN
2. Powered Aircraft On
3. Conducted a normal preflight setup
4. Checked that normal law indications show on PFD (green " on top of the attitude indicator) and that the sidestick indication is available after one engine running.
5. Taxied out and Departed KLAS
6. Check that indications on the lower SD page show valid data.
7. Climbed to cruise
8. Verified no erroneous ecam/PFD messages or GPWS warnings are triggered during flight and descent

Testing Results: 
Passed <!-- Required section --> 

Conclusions: 
Works as intended

Media:
![1250410_20240922203253_1](https://github.com/user-attachments/assets/9445da66-d360-4857-a016-e0681f4a57da)
![1250410_20240922212222_1](https://github.com/user-attachments/assets/47f2ac49-cfd3-4623-b9d2-4bf2a149d1c6)
![1250410_20240922212221_1](https://github.com/user-attachments/assets/80d62885-fd66-4798-bc79-7e8cc3324e9c)
![1250410_20240922212220_1](https://github.com/user-attachments/assets/e1007674-360a-4578-8dc7-61a2706b4d8c)
![1250410_20240922212219_1](https://github.com/user-attachments/assets/77a920be-928d-4294-a63e-a4ac37b2c4ab)
![1250410_20240922212218_1](https://github.com/user-attachments/assets/8116b317-a805-4b28-937e-10051e9b4d89)
![1250410_20240922212216_1](https://github.com/user-attachments/assets/6bc45c2c-5fa6-4981-a497-d95d4a1be4b2)
![1250410_20240922212215_1](https://github.com/user-attachments/assets/ee4e720a-0f52-4bad-ab67-0cbb6201963b)
![1250410_20240922212214_1](https://github.com/user-attachments/assets/efc5c28e-5188-4da8-938f-ab9499439672)
![1250410_20240922212212_1](https://github.com/user-attachments/assets/660749f7-c204-4176-a3c7-53cb183d59c4)
![1250410_20240922212211_1](https://github.com/user-attachments/assets/cffc93fb-f13e-4d22-910e-ff0cb6abc0ed)
![1250410_20240922212210_1](https://github.com/user-attachments/assets/2e029729-6107-4f81-bd00-977450098d50)
